### PR TITLE
Bounds Check and Test Refactor

### PR DESCRIPTION
**Pull request by sdquiring**
Added bounds check mode:
 - When validating all elements, the new client can fill the locations surrounding the input/output data with invalid data, and check that the data was not used/overwritten.

Refactoring pytest:
 - All YAML-only tests are now automatically found and run via the
   single test_config test function.
 - YamlBuilder now constructs documents via an object graph rather than
   via text.
 - All tests that need a build directory can use one specified via the
   command line.

### CODE REVIEW

File: HostLibraryTests/client/DataInitialization_test.cpp [hunk 1]
```diff
@@ -149,23 +205,207 @@ using InputTypes = ::testing::Types<
...
+    EXPECT_EQ(DataInitialization::isBadInput(value), true) << value;
+    EXPECT_EQ(DataInitialization::isBadOutput(value), false) << value;
+
+    value = DataInitialization::getValue<Type>(InitMode::BadOutput);
+    EXPECT_EQ(std::isinf(static_cast<Comparison>(value)), true) << value;
+    EXPECT_EQ(DataInitialization::isBadInput(value), false) << value;
+    EXPECT_EQ(DataInitialization::isBadOutput(value), true) << value;
+}
+
+#if 0
```
Comment (AlexBrownAMD) [inline HostLibraryTests/client/DataInitialization_test.cpp]: Comments

File: HostLibraryTests/client/DataInitialization_test.cpp [hunk 2]
```diff
@@ -149,23 +205,207 @@ using InputTypes = ::testing::Types<
...
+}
+
+TYPED_TEST(DataInitializationTest, Contamination_true_true_true)
 {
-    this->RunDataContaminationTest(false, false);
+    this->RunDataContaminationTest(true, true, true);
 }
 
-TYPED_TEST(DataInitializationTest, Contamination_false_true)
+//TYPED_TEST(DataInitializationTest, BadValues)
```
Comment (AlexBrownAMD) [inline HostLibraryTests/client/DataInitialization_test.cpp]: Comments

File: Tensile/Source/client/source/DataInitialization.cpp [hunk 3]
```diff
@@ -192,6 +203,14 @@ namespace Tensile
                 m_cMaxElements = std::max(m_cMaxElements, problem.c().totalAllocatedElements());
                 m_dMaxElements = std::max(m_dMaxElements, problem.d().totalAllocatedElements());
             }
+
+            if(m_boundsCheck)
+            {
+                m_aMaxElements  += 1024;
```
[the remark this answers is not in the document]
Comment (sdquiring) [reply Tensile/Source/client/source/DataInitialization.cpp]: I'd be open to suggestions for a better/configurable number, but right now we allocate 1024 elements more than the max required for any problem in this run.

**Comment by amdkila** [comment]
Remaining test failures are architecture-specific and occurred before this PR

**Pull request merged**